modules/order/src/AvailabilityOrderProcessor.php

```diff
@@ -37,10 +38,23 @@ public function process(OrderInterface $order) {
     foreach ($order->getItems() as $order_item) {
       $purchased_entity = $order_item->getPurchasedEntity();
       if ($purchased_entity) {
-        $available = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
-        if (!$available) {
-          $order->removeItem($order_item);
-          $order_item->delete();
+        /** @var \Drupal\commerce\AvailabilityResponseInterface $availability */
+        $availability = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
+        if (!$availability->isNeutral()) {
+          if ($availability->getMax() == 0) {
+            $order->removeItem($order_item);
+            $order_item->delete();
+            drupal_set_message(t('The item %item is no longer available and has been removed from your cart.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+          }
+          elseif ($availability->getMax() < $order_item->getQuantity()) {
+            drupal_set_message(new PluralTranslatableMarkup(abs($availability->getMax() - $order_item->getQuantity()), '1 %item is currently unavailable. The quantity in your cart has been updated.', '@count %item are currently unavailable. The quantity in your cart has been updated.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+            $order_item->setQuantity($availability->getMax());
+            $order_item->save();
+          }
```

Comment:
I think you are got a few separate issues (although I can c the link). 
Yesterday me and @steveoliver  covered some of this. I explained the idea behind stock that separated the "is in stock" & the "get total stock level" methods. That was to allow cases like yours. However as the availability checker is passed in the requested quantity, it can simply return that as the max to get around the issue. Steve has shown me its a technique he already used. I will try and address your other points in the drupal issue q as it does not directly relate to the code.